src/watchman_client.js

```diff
@@ -0,0 +1,341 @@
+            watcherInfo.root = root;
+            watcherInfo.relativePath = '';
+            resolve(resp);
+          }
+        });
+      }
+    });
+    
+    return promise;
+  }
+  
+  /**
+   * Issue the 'clock' command to get the time value for use with the 'since'
+   * option during 'subscribe'.
+   */
+  function clock(subscription) {
+    var self = this;
+
+    let promise = new Promise((resolve, reject) => {
+      let watcherInfo = getWatcherInfo(subscription);
+
+      _client.command(['clock', watcherInfo.root], (error, resp) => {
+        if (error) {
+          reject(error);
+        } else {
+          watcherInfo.since = resp.clock;
+          resolve(resp);
+        }
+      });
+    });
+
+    return promise;
+  }
+
+  /**
+   * Called from WatchmanWatcher (or this object during reconnect) to create
+   * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+   * watchman.Client, to be handled in this object.
+   */
+  function subscribe(watchmanWatcher, root) {
+    var self = this;
+
+    let watcherInfo = createWatcherInfo(watchmanWatcher);
+    let subscription = watcherInfo.subscription;
+    
+    let promise = new Promise((resolve, reject) => {
+      watch(subscription, root)
+        .then((resp) => clock(subscription))
+        .then((data) => {
+            let watcherInfo = getWatcherInfo(subscription);
+
+            // create the 'bare' options w/o 'since' or relative_root.
+            // Store in watcherInfo for later use if we need to reset
+            // things after an 'end' caught here.
+            let options = watchmanWatcher.createOptions();  
+            watcherInfo.options = options;
+            
+            // Dup the options object so we can add 'relative_root' and 'since'
+            // and leave the original options object alone. We'll do this again
+            // later if we need to resubscribe after 'end' and reconnect.
+            options = Object.assign({}, options);
+
+            if (_relative_root) {
+              options.relative_root = watcherInfo.relativePath;
+            }
+
+            options.since = watcherInfo.since;
+
+            _client.command(['subscribe',
+                             watcherInfo.root,
+                             subscription,
+                             options],
+                            (error, resp) => {
+                              if (error) {
+                                reject(error);
+                              } else {
+                                resolve(resp);
+                              }
+                            });
+            console.log("For subscription '" + subscription +
+                        "', subscribing against root '" + watcherInfo.root +
+                        "', with relative_root = '" + options.relative_root + "'");
+        })
+        .catch(function(error) {
+          console.error("Caught error in subscribe for watchmanWatcher root " + root);
```

Comment:
Unless there is a really strong motivation, I would prefer if libraries don't log things  by default. Otherwise log output gets unwieldy.

If there is strong motivation let me know, otherwise I suspect we may not want to alwayas log this, rather using something like https://www.npmjs.com/package/debug

Reply:
Stef, the error indicates that a subscription failed. Isn't that something that should be reported? (In general I agree with moving most things to debug, but not sure on this one).

Reply:
Comments on my comment? I could see putting something in WatchmanWatcher and removing it from WatchmanClient, but I do think that swallowing the error (by just rejecting and not putting anything out at all) isn't a great solution. Comments?